docs/specs/user-profile/design.md

```diff
@@ -0,0 +1,329 @@
+
+#### Access control on custom attributes for Admin API and the portal
+
+The Admin API and the portal always have full access to custom attributes.
+
+#### Access control on custom attributes for the resolver
+
+The resolver first checks if the session has client ID.
+If a client ID is present, then the access control of the client application is applied.
+If a client ID is absent, then the full custom attributes are included in the response header.
+
+#### Access control on custom attributes for client application
+
+Client application can access the custom attributes via the User Info endpoint.
+Client application does NOT have write access to custom attributes because the User Info endpoint is not a mutation.
+Client application by default has read access to all custom attributes via the User Info endpoint.
+The developer can restrict access by declaring which custom attribute the client application has access to.
+
+```yaml
+oauth:
+  clients:
+  - client_id: my_client
+    allowed_custom_attributes:
+    # Other custom attributes not listed here are not visible to this client.
+    # Each entry is a JSON pointer in JSON string representation.
+    - /app_user_role
+```
+
+#### Access control on custom attributes for end-user
+
+The end-user can interact with custom attributes via 2 means.
+
+Depending on which client application is being used, the end-user can see different set of custom attributes.
+Thus the access control for client application also affects the end-user.
+
+Custom attributes can also be configured to be read-only or read-write in the settings page.
+Undeclared custom attributes are hidden in the settings page.
+
+```yaml
+ui:
+  custom_attributes:
+  - pointer: /app_user_role
+    access_control: readonly
+  - pointer: /hobby
+    access_control: readwrite
+```
+
+### Editing custom attributes in the settings page and in the portal
+
+- Custom attributes without `type` are ignored.
+- Custom attributes of `type` `boolean` is a `<input type="checkbox">`
+- Custom attributes of `type` `string` is a `<input type="text">`
+- Custom attributes of `type` `number` is a `<input type="text">`
+- Custom attributes of `type` `integer` is a `<input type="text">`
+
+## ID Token
+
+The ID token never contain standard attributes nor custom attributes.
+This is because ID token can be used as `id_token_hint` and can appear in URL query.
+
+## User Info endpoint
+
+The standard attributes appear in the root of the user info response.
+The custom attributes appear under the key `custom_attributes`.
+
+Here is an example of the response.
+
+```json
+{
+  "sub": "user_id",
+  "email": "user@example.com",
+  "email_verified": true,
+  "given_name": "John",
+  "family_name": "Doe",
+  "custom_attributes": {
+    "hobby": "reading"
+  }
+}
+```
+
+## Synchronization of user profile between Authgear and the backend server
+
+There is NO blocking nor non-blocking webhook when standard attributes and custom attributes
```

Comment:
As we discussed in meeting last time, what about if developer want to add server side validation? (Not a must in 1st implementation)

Reply:
and say in the case of `zoneinfo` or `locale`, have a webhook to know it is changed?

Reply:
> As we discussed in meeting last time, what about if developer want to add server side validation? (Not a must in 1st implementation)

I would recommend they store that information themselves.

> and say in the case of zoneinfo or locale, have a webhook to know it is changed?

As I mentioned in the use case examples, if the developer wants to know when the value is changed,
they probably need that information for their business logic.
In that case I highly recommend they store zoneinfo and locale themselves,
and they tell us the new values so that localization of Authgear is consistent with their application.

Reply:
I think assuming all developers would not expect using Authgear custom attr is not the single source of truth is not realistic